File: src/main/java/com/zepben/vertxutils/json/LazyJsonObject.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.vertxutils.json;

import com.zepben.annotations.EverythingIsNonnullByDefault;
import io.vertx.core.buffer.Buffer;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import org.apache.commons.lang3.ObjectUtils;

import javax.annotation.Nullable;
import java.time.Instant;
import java.util.*;
import java.util.function.Consumer;
import java.util.function.Supplier;
import java.util.stream.Stream;

@EverythingIsNonnullByDefault
@SuppressWarnings({"WeakerAccess"})
public class LazyJsonObject extends JsonObject {

    @Nullable
    private Supplier<JsonObject> jsonObjectSupplier;
    private boolean materialised = false;
    private final Map<String, Supplier<?>> fieldSuppliers = new HashMap<>();

    public LazyJsonObject(Supplier<JsonObject> supplier) {
        this.jsonObjectSupplier = supplier;
    }

    public LazyJsonObject() {
    }

    @Override
    public <T> T mapTo(Class<T> type) {
        checkObjectMaterialised();
        checkFieldsMaterialised();
        return super.mapTo(type);
    }

    @Override
    public String getString(String key) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getString(key);
    }

    @Override
    public Integer getInteger(String key) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getInteger(key);
    }

    @Override
    public Long getLong(String key) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getLong(key);
    }

    @Override
    public Double getDouble(String key) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getDouble(key);
    }

    @Override
    public Float getFloat(String key) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getFloat(key);
    }

    @Override
    public Boolean getBoolean(String key) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getBoolean(key);
    }

    @Override
    public JsonObject getJsonObject(String key) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getJsonObject(key);
    }

    @Override
    public JsonArray getJsonArray(String key) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getJsonArray(key);
    }

    @Override
    public byte[] getBinary(String key) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getBinary(key);
    }

    @Override
    public Instant getInstant(String key) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getInstant(key);
    }

    @Override
    public Object getValue(String key) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getValue(key);
    }

    @Override
    public String getString(String key, String def) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getString(key, def);
    }

    @Override
    public Integer getInteger(String key, Integer def) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getInteger(key, def);
    }

    @Override
    public Long getLong(String key, Long def) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getLong(key, def);
    }

    @Override
    public Double getDouble(String key, Double def) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getDouble(key, def);
    }

    @Override
    public Float getFloat(String key, Float def) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getFloat(key, def);
    }

    @Override
    public Boolean getBoolean(String key, Boolean def) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getBoolean(key, def);
    }

    @Override
    public JsonObject getJsonObject(String key, JsonObject def) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getJsonObject(key, def);
    }

    @Override
    public JsonArray getJsonArray(String key, JsonArray def) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getJsonArray(key, def);
    }

    @Override
    public byte[] getBinary(String key, byte[] def) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getBinary(key, def);
    }

    @Override
    public Instant getInstant(String key, Instant def) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getInstant(key, def);
    }

    @Override
    public Object getValue(String key, Object def) {
        checkObjectMaterialised();
        checkFieldMaterialised(key);
        return super.getValue(key, def);
    }

    @Override
    public boolean containsKey(String key) {
        checkObjectMaterialised();
        return super.containsKey(key) || fieldSuppliers.containsKey(key);
    }

    @Override
    public Set<String> fieldNames() {
        checkObjectMaterialised();
        Set<String> names = new HashSet<>(super.fieldNames());
        names.addAll(fieldSuppliers.keySet());
        return names;
    }

    // Warning: This may return an instance of Supplier
    @Override
    public Object remove(String key) {
        checkObjectMaterialised();
        return ObjectUtils.firstNonNull(fieldSuppliers.remove(key), super.remove(key));
    }

    // Note: if other is a LazyJsonObject, mergeIn will replace Suppliers rather than recursively merging them
    @Override
    public JsonObject mergeIn(JsonObject other, int depth) {
        if (depth < 1) {
            return this;
        }

        checkObjectMaterialised();

        if (other instanceof LazyJsonObject) {
            LazyJsonObject ljo = (LazyJsonObject) other;
            for (String f : other.fieldNames()) {
                super.remove(f);
                fieldSuppliers.remove(f);
            }
            fieldSuppliers.putAll(ljo.fieldSuppliers);
        }

        return super.mergeIn(other, depth);
    }

    @Override
    public String encode() {
        checkObjectMaterialised();
        checkFieldsMaterialised();
        return super.encode();
    }

    @Override
    public String encodePrettily() {
        checkObjectMaterialised();
        checkFieldsMaterialised();
        return super.encodePrettily();
    }

    @Override
    public Buffer toBuffer() {
        checkObjectMaterialised();
        checkFieldsMaterialised();
        return super.toBuffer();
    }

    @Override
    public JsonObject copy() {
        checkObjectMaterialised();
        checkFieldsMaterialised();
        return super.copy();
    }

    @Override
    public Map<String, Object> getMap() {
        checkObjectMaterialised();
        checkFieldsMaterialised();
        return super.getMap();
    }

    @Override
    public Stream<Map.Entry<String, Object>> stream() {
        checkObjectMaterialised();
        checkFieldsMaterialised();
        return super.stream();
    }

    @Override
    public Iterator<Map.Entry<String, Object>> iterator() {
        checkObjectMaterialised();
        checkFieldsMaterialised();
        return super.iterator();
    }

    @Override
    public int size() {
        checkObjectMaterialised();
        return super.size() + fieldSuppliers.size();
    }

    @Override
    public boolean isEmpty() {
        checkObjectMaterialised();
        return super.isEmpty() && fieldSuppliers.isEmpty();
    }

    @Override
    public String toString() {
        checkObjectMaterialised();
        checkFieldsMaterialised();
        return super.toString();
    }

    @Override
    public boolean equals(Object o) {
        checkObjectMaterialised();
        checkFieldsMaterialised();
        return super.equals(o);
    }

    @Override
    public int hashCode() {
        checkObjectMaterialised();
        return super.hashCode() + fieldSuppliers.hashCode();
    }

    @Override
    public void writeToBuffer(Buffer buffer) {
        checkObjectMaterialised();
        checkFieldsMaterialised();
        super.writeToBuffer(buffer);
    }

    @Override
    public void forEach(Consumer<? super Map.Entry<String, Object>> action) {
        checkObjectMaterialised();
        checkFieldsMaterialised();
        super.forEach(action);
    }

    @Override
    public Spliterator<Map.Entry<String, Object>> spliterator() {
        checkObjectMaterialised();
        checkFieldsMaterialised();
        return super.spliterator();
    }

    public void put(String key, Supplier<?> supplier) {
        fieldSuppliers.put(key, supplier);
    }

    public LazyJsonObject lazyPut(String key, Supplier<?> supplier) {
        fieldSuppliers.put(key, supplier);
        return this;
    }

    private void checkObjectMaterialised() {
        if (!materialised && jsonObjectSupplier != null) {
            materialised = true;
            mergeIn(jsonObjectSupplier.get());
        }
    }

    private void checkFieldsMaterialised() {
        fieldSuppliers.forEach((key, supplier) -> super.put(key, supplier.get()));
        fieldSuppliers.clear();
    }

    private void checkFieldMaterialised(String key) {
        Supplier<?> supplier = fieldSuppliers.remove(key);
        if (supplier != null) {
            super.put(key, supplier.get());
        }
    }

}

File: src/main/java/com/zepben/vertxutils/json/LazyJsonObject.kt
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.vertxutils.json

import io.vertx.core.buffer.Buffer
import io.vertx.core.json.JsonArray
import io.vertx.core.json.JsonObject
import java.time.Instant
import java.util.*
import java.util.function.Consumer
import java.util.stream.Stream

class LazyJsonObject(
    private val jsonObjectSupplier: (() -> JsonObject)? = null,
) : JsonObject() {

    private var materialised = false
    private val fieldSuppliers = mutableMapOf<String, () -> Any?>()

    override fun <T> mapTo(type: Class<T?>): T? {
        checkObjectMaterialised()
        checkFieldsMaterialised()
        return super.mapTo(type)
    }

    override fun getString(key: String): String? {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getString(key)
    }

    override fun getInteger(key: String): Int? {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getInteger(key)
    }

    override fun getLong(key: String): Long? {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getLong(key)
    }

    override fun getDouble(key: String): Double? {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getDouble(key)
    }

    override fun getFloat(key: String): Float? {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getFloat(key)
    }

    override fun getBoolean(key: String): Boolean? {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getBoolean(key)
    }

    override fun getJsonObject(key: String): JsonObject? {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getJsonObject(key)
    }

    override fun getJsonArray(key: String): JsonArray? {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getJsonArray(key)
    }

    override fun getBinary(key: String): ByteArray? {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getBinary(key)
    }

    override fun getInstant(key: String): Instant? {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getInstant(key)
    }

    override fun getValue(key: String): Any? {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getValue(key)
    }

    override fun getString(key: String, def: String): String {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getString(key, def)
    }

    override fun getInteger(key: String, def: Int): Int {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getInteger(key, def)
    }

    override fun getLong(key: String, def: Long): Long {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getLong(key, def)
    }

    override fun getDouble(key: String, def: Double): Double {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getDouble(key, def)
    }

    override fun getFloat(key: String, def: Float): Float {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getFloat(key, def)
    }

    override fun getBoolean(key: String, def: Boolean): Boolean {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getBoolean(key, def)
    }

    override fun getJsonObject(key: String, def: JsonObject): JsonObject {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getJsonObject(key, def)
    }

    override fun getJsonArray(key: String, def: JsonArray): JsonArray {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getJsonArray(key, def)
    }

    override fun getBinary(key: String, def: ByteArray): ByteArray {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getBinary(key, def)
    }

    override fun getInstant(key: String, def: Instant): Instant {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getInstant(key, def)
    }

    override fun getValue(key: String, def: Any): Any {
        checkObjectMaterialised()
        checkFieldMaterialised(key)
        return super.getValue(key, def)
    }

    override fun containsKey(key: String): Boolean {
        checkObjectMaterialised()
        return super.containsKey(key) || fieldSuppliers.containsKey(key)
    }

    override fun fieldNames(): Set<String> {
        checkObjectMaterialised()
        return super.fieldNames() + fieldSuppliers.keys
    }

    // Warning: This may return an instance of Supplier
    override fun remove(key: String): Any? {
        checkObjectMaterialised()
        return sequenceOf(fieldSuppliers.remove(key), super.remove(key)).filterNotNull().firstOrNull()
    }

    // Note: if other is a LazyJsonObject, mergeIn will replace Suppliers rather than recursively merging them
    override fun mergeIn(other: JsonObject, depth: Int): JsonObject {
        if (depth < 1) {
            return this
        }

        checkObjectMaterialised()

        if (other is LazyJsonObject) {
            val ljo = other
            other.fieldNames().forEach { f ->
                super.remove(f)
                fieldSuppliers.remove(f)
            }
            fieldSuppliers.putAll(ljo.fieldSuppliers)
        }

        return super.mergeIn(other, depth)
    }

    override fun encode(): String {
        checkObjectMaterialised()
        checkFieldsMaterialised()
        return super.encode()
    }

    override fun encodePrettily(): String {
        checkObjectMaterialised()
        checkFieldsMaterialised()
        return super.encodePrettily()
    }

    override fun toBuffer(): Buffer {
        checkObjectMaterialised()
        checkFieldsMaterialised()
        return super.toBuffer()
    }

    override fun copy(): JsonObject {
        checkObjectMaterialised()
        checkFieldsMaterialised()
        return super.copy()
    }

    override fun getMap(): MutableMap<String, Any?> {
        checkObjectMaterialised()
        checkFieldsMaterialised()
        return super.getMap()
    }

    override fun stream(): Stream<MutableMap.MutableEntry<String, Any>> {
        checkObjectMaterialised()
        checkFieldsMaterialised()
        return super.stream()
    }

    override fun iterator(): MutableIterator<MutableMap.MutableEntry<String, Any>> {
        checkObjectMaterialised()
        checkFieldsMaterialised()
        return super.iterator()
    }

    override fun size(): Int {
        checkObjectMaterialised()
        return super.size() + fieldSuppliers.size
    }

    override fun isEmpty(): Boolean {
        checkObjectMaterialised()
        return super.isEmpty() && fieldSuppliers.isEmpty()
    }

    override fun toString(): String {
        checkObjectMaterialised()
        checkFieldsMaterialised()
        return super.toString()
    }

    override fun equals(o: Any?): Boolean {
        checkObjectMaterialised()
        checkFieldsMaterialised()
        return super.equals(o)
    }

    override fun hashCode(): Int {
        checkObjectMaterialised()
        return super.hashCode() + fieldSuppliers.hashCode()
    }

    override fun writeToBuffer(buffer: Buffer) {
        checkObjectMaterialised()
        checkFieldsMaterialised()
        super.writeToBuffer(buffer)
    }

    override fun forEach(action: Consumer<in MutableMap.MutableEntry<String, Any?>>) {
        checkObjectMaterialised()
        checkFieldsMaterialised()
        super.forEach(action)
    }

    fun forEach(action: (MutableMap.MutableEntry<String, Any?>) -> Unit) {
        checkObjectMaterialised()
        checkFieldsMaterialised()
        super.forEach(action)
    }

    override fun spliterator(): Spliterator<MutableMap.MutableEntry<String, Any?>> {
        checkObjectMaterialised()
        checkFieldsMaterialised()
        return super.spliterator()
    }

    operator fun set(key: String, supplier: () -> Any?) {
        fieldSuppliers[key] = supplier
    }

    fun put(key: String, supplier: () -> Any?) {
        fieldSuppliers[key] = supplier
    }

    fun lazyPut(key: String, supplier: () -> Any?): LazyJsonObject {
        fieldSuppliers[key] = supplier
        return this
    }

    private fun checkObjectMaterialised() {
        if (!materialised && (jsonObjectSupplier != null)) {
            materialised = true
            mergeIn(jsonObjectSupplier())
        }
    }

    private fun checkFieldsMaterialised() {
        fieldSuppliers.forEach { (key, supplier) -> super.put(key, supplier()) }
        fieldSuppliers.clear()
    }

    private fun checkFieldMaterialised(key: String) {
        fieldSuppliers.remove(key)?.also { supplier -> super.put(key, supplier()) }
    }

}
